### PR TITLE
Move RowExpression translation above AddExchange

### PR DESCRIPTION
**Pull request by highker**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by highker** [comment]
Ready for review. @rongrong, @wenleix, @arhimondr, just noticed that it is safe to move connector optimization rule before `AddExchange`. In that way, we only need "logical optimization" for connectors. That will serve as the only interface to let connectors to absorb part of a plan. cc: @agrawaldevesh

**Comment by highker** [comment]
NVM, we might still need two.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 1]
```diff
@@ -1396,8 +1392,8 @@ private boolean canPushdownPartialMergeThroughLowMemoryOperators(PlanNode node)
     {
         Map<VariableReferenceExpression, VariableReferenceExpression> outputToInput = new HashMap<>();
         for (Map.Entry<VariableReferenceExpression, RowExpression> assignment : assignments.getMap().entrySet()) {
```
Comment (hellium01) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: Remove unused parameter `TypeProvider types`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java [hunk 2]
```diff
@@ -424,18 +561,78 @@ private boolean isCandidate(Map<ColumnHandle, NullableValue> bindings)
...
+            return !Boolean.FALSE.equals(optimized) && optimized != null && (!(optimized instanceof ConstantExpression) || !((ConstantExpression) optimized).isNull());
+        }
+    }
+
+    private static class LookupVariableResolver
+            implements VariableResolver
+    {
+        private final Map<VariableReferenceExpression, ColumnHandle> assignments;
+        private final Map<ColumnHandle, NullableValue> bindings;
+        private final Function<VariableReferenceExpression, Object> missingBindingSupplier;
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java]: Add a comment why the value type is `Object`, and maybe a TODO stating change to `RowExpression` once all migration is done? 

Is it possible to make `LookupVariableResolver` a generic class?

**Pull request merged**